src/HDWallet.cpp

```
@@ -89,102 +80,60 @@ HDWallet<seedSize>::~HDWallet() {
 }
 
-template <std::size_t seedSize>
-DerivationPath HDWallet<seedSize>::cardanoStakingDerivationPath(const DerivationPath& path) {
-    DerivationPath stakingPath = path;
-    stakingPath.indices[3].value = 2;
-    stakingPath.indices[4].value = 0;
-    return stakingPath;
-}
-
 template <std::size_t seedSize>
 PrivateKey HDWallet<seedSize>::getKeyByCurve(TWCurve curve, const DerivationPath& derivationPath) const {
-    const auto privateKeyType = PrivateKey::getType(curve);
-    auto node = getNode<seedSize>(*this, curve, derivationPath);
-    switch (privateKeyType) {
-    case TWPrivateKeyTypeCardano: {
-        if (derivationPath.indices.size() < 4 || derivationPath.indices[3].value > 1) {
-            // invalid derivation path
-            return PrivateKey(Data(PrivateKey::cardanoKeySize), curve);
-        }
-        const DerivationPath stakingPath = cardanoStakingDerivationPath(derivationPath);
-
-        auto pkData = Data(node.private_key, node.private_key + PrivateKey::_size);
-        auto extData = Data(node.private_key_extension, node.private_key_extension + PrivateKey::_size);
-        auto chainCode = Data(node.chain_code, node.chain_code + PrivateKey::_size);
-
-        // repeat with staking path
-        const auto node2 = getNode(*this, curve, stakingPath);
-        auto pkData2 = Data(node2.private_key, node2.private_key + PrivateKey::_size);
-        auto extData2 = Data(node2.private_key_extension, node2.private_key_extension + PrivateKey::_size);
-        auto chainCode2 = Data(node2.chain_code, node2.chain_code + PrivateKey::_size);
-
-        TW::memzero(&node);
-        return PrivateKey(pkData, extData, chainCode, pkData2, extData2, chainCode2, curve);
+    auto curveToUse = curve == TWCurveStarkex ? TWCurveSECP256k1 : curve;
+    auto node = getNode<seedSize>(*this, curveToUse, derivationPath, Hash::HasherSha256ripemd);
+    if (node == nullptr) {
+        throw std::invalid_argument("Invalid node");
```

Let's throw `Invalid derivation path` maybe?